cmd/clusterctl/client/upgrade.go

```diff
@@ -30,6 +30,8 @@ import (
 	"sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster"
 )
 
+const upgradeItemProviderNameError = "invalid provider name %q. Provider name should be in the form namespace/provider[:version] or provider:version"
```

Comment:
also, should we change the description for flags and for fields in the ApplyUpgradeOptions struct describing that we now support provider:version? we can eventually push things a little bit further by stating that the old format is deprecated and will be removed in a follow-up release

Reply:
I think provider[:version] would be an improvement. It wasn't implemented since it was explicitly mentioned in other places [`"invalid provider name %q. Provider name should be in the form name[:version] and version cannot be empty"`](https://github.com/kubernetes-sigs/cluster-api/blob/c4844f215954e813dfb2f69f94736563e7b40ad3/cmd/clusterctl/client/common.go#L81). I wasn't sure if it would cause any other issues so I just focused on removing namespace.

Q: Would you like to remove version and if so, in this pr?

Updating the fields in ApplyUpgradeOptions sounds great. I will fix it.

Reply:
I do expect provider[:version] should resolve in latest, is this correct?

Reply:
@fabriziopandini I tested it with almost current clusterctl. 

Current clusterctl without version
```bash
clusterctl upgrade apply --infrastructure capd-system/docker
Checking cert-manager version...
Cert-manager is already up to date
Error: invalid provider name "capd-system/docker". Provider name should be in the form namespace/name:version and version cannot be empty
```

Current clusterctl with version
```bash
oscaru@Loke:~/go/src/k8s.io/kubernetes-sig/cluster-api$ clusterctl upgrade apply --infrastructure capd-system/docker:v1.2.5
Checking cert-manager version...
Cert-manager is already up to date
Performing upgrade...
Scaling down Provider="infrastructure-docker" Version="" Namespace="capd-system"
Deleting Provider="infrastructure-docker" Version="" Namespace="capd-system"
Installing Provider="infrastructure-docker" Version="v1.2.5" TargetNamespace="capd-system"

New clusterctl version available: v1.2.2 -> v1.2.5
https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.2.5
```

I would gladly also remove version as required. I suggest in a follow up pr but I can do it in this one if you prefer.

Reply:
+1 to do one step at time.
If version is required with the old syntax (Provider name should be in the form namespace/name:version and version cannot be empty" as it is stated in https://github.com/kubernetes-sigs/cluster-api/blob/aacbbc780a42363b28c232cccc87bfecd816df0b/cmd/clusterctl/client/upgrade.go#L220 (which seems to me to override to the error message in https://github.com/kubernetes-sigs/cluster-api/blob/aacbbc780a42363b28c232cccc87bfecd816df0b/cmd/clusterctl/client/upgrade.go#L227) , the new syntax should just drop namespace with this PR, but version should remain mandatory

Then we can eventually discuss making version optional/defaulting to the latest available version in a follow up PR, but that's TBD and somehow also against our current recommendation to pin version 

> Even if slightly verbose, pinning the version provides a better control over what is installed, as usually required in an enterprise environment, especially if you rely on an internal repository with a separated software supply chain or a custom versioning schema.

Reply:
Great! Then I will rebase and clean this pr as soon as 1.3 is released. If it's decided to make version optional I can do that as soon as this pr merges. I've just been in this part of the code so hopefully it will be straight forward.

Reply:
Let's get this merged first, I'm not really sold to the idea of making the version optional because I think that our recommendation to pin the version is correct and our UX should enforce this

Reply:
I think if optional version is careful: only selecting latest patch version and only informing the user about later major / minor versions so they can do it themselves (after reading the upgrade guide) it might be "nice to have"